### PR TITLE
[MGPG-125] Fix "bestPractices"

### PR DESCRIPTION
**Pull request by cstamas**
As currently due defaulted values it would always fail, unless user explicitly re-configures plugin. So, do not set "default value" on mojo level, instead handle it programmatically, when bestPractices is not set. Added ITs for "bestPractices", one for success and one for failure. 

---

https://issues.apache.org/jira/browse/MGPG-125

### CODE REVIEW

File: src/it/sign-release-best-practices-fail/verify.groovy [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+
+import org.codehaus.plexus.util.FileUtils
+
+var buildLog = new File(basedir, "build.log")
+var logContent = FileUtils.fileRead(buildLog)
```
Comment (bmarwell) [inline src/it/sign-release-best-practices-fail/verify.groovy:24]: Groovy can do that without plexus utils:

```
String logContent = new File( basedir, "build.log").text
```

File: src/it/sign-release-best-practices-fail/verify.groovy [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+// assert that bestPractice+worstPractice => MojoFailure
+if (!logContent.contains("MojoFailureException: Do not store passphrase in any file")) {
+    throw new Exception("Maven build did not fail in consequence of bestPractices+worstPractices")
+}
```
Comment (bmarwell) [inline src/it/sign-release-best-practices-fail/verify.groovy:29]: Groovy's native assert is quite suitable here:

```
assert logContent.contains( "..." )
```

File: src/main/java/org/apache/maven/plugins/gpg/AbstractGpgMojo.java [hunk 3]
```diff
@@ -299,11 +300,18 @@ public final void execute() throws MojoExecutionException, MojoFailureException
...
-                    "Do not store passphrase in any file (disk or SCM repository), rely on GnuPG agent or provide passphrase in "
-                            + passphraseEnvName + " environment variable.");
+        if (bestPractices) {
+            if (isNotBlank(passphrase) || isNotBlank(passphraseServerId)) {
+                // Stop propagating worst practices: passphrase MUST NOT be in any file on disk
+                throw new MojoFailureException(
+                        "Do not store passphrase in any file (disk or SCM repository), rely on GnuPG agent or provide passphrase in "
+                                + passphraseEnvName + " environment variable.");
+            }
+        } else {
```
Comment (bmarwell) [inline src/main/java/org/apache/maven/plugins/gpg/AbstractGpgMojo.java:310]: As best practices will probably be extended in the future, why not just refactor out a method here?

```suggestion
        if (bestPractices) {
            checkForBestPractices();
        } else {
```

File: src/main/java/org/apache/maven/plugins/gpg/AbstractGpgMojo.java [hunk 4]
```diff
@@ -299,11 +300,18 @@ public final void execute() throws MojoExecutionException, MojoFailureException
...
+            if (!isNotBlank(passphraseServerId)) {
+                // default it programmatically: this is needed to handle different cases re bestPractices
+                passphraseServerId = GPG_PASSPHRASE;
+            }
```
Comment (bmarwell) [inline src/main/java/org/apache/maven/plugins/gpg/AbstractGpgMojo.java:314]: I feel this is a default case inside an else-if nesting. However, as the isolated if(bestPractices) improves readability, I'd say: keep it for now.

One solution would be to refactor out a method like `setDefaults()`, but probably not worth it.

**Review by bmarwell — COMMENTED** [review]
Mostly a little groovy and nesting readability thoughts.

File: src/main/java/org/apache/maven/plugins/gpg/AbstractGpgMojo.java (hunk 4 above)
Comment (cstamas) [reply src/main/java/org/apache/maven/plugins/gpg/AbstractGpgMojo.java:314]: These fields are deprecated and are meant to be removed.. so IMO that would be overkill.

**Comment by cstamas** [comment]
Btw, this whole PR (with all "non groovy-ness") was copy pasted from other ITs... So maybe we need a "sweeping change" instead where we rewrite the ITs in this plugin to "groovy-like" constructs.

File: src/it/sign-release-best-practices-fail/pom.xml [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+  <properties>
+    <maven.test.skip>true</maven.test.skip>
+  </properties>
+
+  <build>
+    <plugins>
+      <plugin>
+        <groupId>org.apache.maven.plugins</groupId>
+        <artifactId>maven-compiler-plugin</artifactId>
+        <version>2.0.2</version>
```
Comment (slawekjaranowski) [inline src/it/sign-release-best-practices-fail/pom.xml:43]: ?
we can use `@version.maven-compiler-plugin@` properties for plugins versions are defined in parent

File: src/it/sign-release-best-practices-fail/pom.xml [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+          <bestPractices>true</bestPractices>
+          <!-- load passphrase from settings.xml -->
+          <!-- but not the default -->
+          <passphraseServerId>gpg.passphrase-sign-with-passphase-from-maven-settings</passphraseServerId>
+        </configuration>
+      </plugin>
+      <plugin>
+        <groupId>org.apache.maven.plugins</groupId>
+        <artifactId>maven-install-plugin</artifactId>
+        <version>2.2</version>
```
Comment (slawekjaranowski) [inline src/it/sign-release-best-practices-fail/pom.xml:67]: 😄

File: src/it/sign-release-best-practices/verify.groovy [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+    "test-1.0-sources.jar",
+    "test-1.0-sources.jar.asc"
+]
+
+for (String expectedFile : expectedFiles) {
+    var file = new File(artifactDir, expectedFile)
+
+    println "Checking for existence of $file"
+
+    if (!file.isFile()) {
```
Comment (slawekjaranowski) [inline src/it/sign-release-best-practices/verify.groovy:35]: `assert file.isFile()`

File: src/main/java/org/apache/maven/plugins/gpg/AbstractGpgMojo.java [hunk 8]
```diff
@@ -131,13 +131,14 @@ public abstract class AbstractGpgMojo extends AbstractMojo {
...
      * If parameter {@link #bestPractices} set to {@code true}, plugin fails when this parameter is configured.</em>
+     * Is programatically defaulted to {@link #GPG_PASSPHRASE}.
      *
      * @since 1.6
      * @deprecated Do not use this configuration, it may leak sensitive information. Rely on gpg-agent or env
      * variables instead.
      **/
     @Deprecated
-    @Parameter(property = "gpg.passphraseServerId", defaultValue = GPG_PASSPHRASE)
+    @Parameter(property = "gpg.passphraseServerId")
```
Comment (slawekjaranowski) [inline src/main/java/org/apache/maven/plugins/gpg/AbstractGpgMojo.java:141]: I would like to not remove default values .... can be assumed as present by many configurations

**Comment by cstamas** [comment]
Updated PR with comments, but now the new IT is considerably slower (?) and also locally did not, but on CI fails (I guess the force sources creation is needed as there are actually no sources...) @slawekjaranowski ?

**Review by slawekjaranowski — APPROVED** [review]

**Comment by cstamas** [comment]
For last build this PR:
```
[INFO] Building: sign-release-best-practices/pom.xml
[INFO] run post-build script verify.groovy
[INFO]           sign-release-best-practices/pom.xml .............. SUCCESS (4.002 s)
```
2nd build/commit:
```
[INFO] Building: sign-release-best-practices/pom.xml
[INFO] run post-build script verify.groovy
[INFO]           sign-release-best-practices/pom.xml .............. SUCCESS (3.877 s)
```

Hm, minimal... unsure what I "feel" locally. Maybe that these two ITs use _new_ plugins (hence downoads them) while all the rest use old/ancient plugins but some version?

**Comment by slawekjaranowski** [comment]
It can be a time needed for downloading new plugins version into tarted/local-repo
Too small difference to more investigating

build with `mvn clean ...`
```
gpg
[INFO]           sign-release-best-practices/pom.xml .............. SUCCESS (8.135 s)

bc
[INFO]           sign-release-best-practices/pom.xml .............. SUCCESS (2.189 s)
```

next without `clean`

```
gpg
[INFO]           sign-release-best-practices/pom.xml .............. SUCCESS (2.653 s)

bc
[INFO]           sign-release-best-practices/pom.xml .............. SUCCESS (2.237 s)

```

**Comment by jira-importer** [comment]
Resolve #257

**Comment by jira-importer** [comment]
Resolve #257

**Pull request merged**